File: MiniDOMBase/DMD/Luoghi/Nazioni/CNazione.cs
using System;
using System.Collections;
using DMD;
using DMD.XML;
using DMD.Databases;
using DMD.Databases.Collections;
using static minidom.Sistema;
using static minidom.Anagrafica;
using minidom.repositories;
using System.Collections.Generic;


namespace minidom
{
    public partial class Anagrafica
    {

        /// <summary>
        /// Nazione
        /// </summary>
        [Serializable]
        public class CNazione 
            : LuogoISTAT
        {
            private string m_SantoPatrono;
            private string m_GiornoFestivo;
            private string m_Prefisso;
            private string m_Sigla;
            

            /// <summary>
            /// Costruttore
            /// </summary>
            public CNazione()
            {
                this.m_SantoPatrono = "";
                this.m_GiornoFestivo = "";
                this.m_Prefisso = "";
                this.m_Sigla = "";
            }

            /// <summary>
            /// Costruttore
            /// </summary>
            /// <param name="nome"></param>
            public CNazione(string nome)
                : base(nome)
            {
            }

            /// <summary>
            /// Restituisce il codice hash dell'oggetto
            /// </summary>
            /// <returns></returns>
            public override int GetHashCode()
            {
                return base.GetHashCode();
            }

            /// <summary>
            /// Restituisce true se i due oggetti sono uguali
            /// </summary>
            /// <param name="obj"></param>
            /// <returns></returns>
            public sealed override bool Equals(LuogoISTAT obj)
            {
                return (obj is CNazione) && this.Equals((CNazione)obj);
            }

            /// <summary>
            /// Restituisce true se i due oggetti sono uguali
            /// </summary>
            /// <param name="obj"></param>
            /// <returns></returns>
            public virtual bool Equals(CNazione obj)
            {
                return base.Equals(obj)
                    && DMD.Strings.EQ(this.m_SantoPatrono, obj.m_SantoPatrono)
                    && DMD.Strings.EQ(this.m_GiornoFestivo, obj.m_GiornoFestivo)
                    && DMD.Strings.EQ(this.m_Prefisso, obj.m_Prefisso)
                    && DMD.Strings.EQ(this.m_Sigla, obj.m_Sigla)
                    ;
            }

            /// <summary>
            /// Restituisce o imposta il nome del santo patrono
            /// </summary>
            public string SantoPatrono
            {
                get
                {
                    return m_SantoPatrono;
                }

                set
                {
                    value = DMD.Strings.Trim(value);
                    string oldValue = m_SantoPatrono;
                    if ((oldValue ?? "") == (value ?? ""))
                        return;
                    m_SantoPatrono = value;
                    DoChanged("SantoPatrono", value, oldValue);
                }
            }

            /// <summary>
            /// Restituisce o imposta una stringa che descrive la data del giorno festivo per il santo patrono
            /// </summary>
            public string GiornoFestivo
            {
                get
                {
                    return m_GiornoFestivo;
                }

                set
                {
                    value = DMD.Strings.Trim(value);
                    string oldValue = m_GiornoFestivo;
                    if ((oldValue ?? "") == (value ?? ""))
                        return;
                    m_GiornoFestivo = value;
                    DoChanged("GiornoFestivo", value, oldValue);
                }
            }

            /// <summary>
            /// Restituisce o imposta il prefisso telefonico principale
            /// </summary>
            public string Prefisso
            {
                get
                {
                    return m_Prefisso;
                }

                set
                {
                    value = Sistema.Formats.ParsePhoneNumber(value);
                    string oldValue = m_Prefisso;
                    if ((oldValue ?? "") == (value ?? ""))
                        return;
                    m_Prefisso = value;
                    DoChanged("Prefisso", value, oldValue);
                }
            }

            /// <summary>
            /// Restituisce o imposta la sigla della nazione
            /// </summary>
            public string Sigla
            {
                get
                {
                    return m_Sigla;
                }

                set
                {
                    value = DMD.Strings.Trim(value);
                    string oldValue = m_Sigla;
                    if ((oldValue ?? "") == (value ?? ""))
                        return;
                    m_Sigla = value;
                    DoChanged("Sigla", value, oldValue);
                }
            }

            //protected internal override Databases.CDBConnection GetConnection()
            //{
            //    return Luoghi.Database;
            //}

            /// <summary>
            /// Repository
            /// </summary>
            /// <returns></returns>
            public override CModulesClass GetModule()
            {
                return minidom.Anagrafica.Luoghi.Nazioni; //.Module;
            }

            /// <summary>
            /// Discriminator
            /// </summary>
            /// <returns></returns>
            public override string GetTableName()
            {
                return "tbl_Luoghi_Nazioni";
            }

            /// <summary>
            /// Carica dal db
            /// </summary>
            /// <param name="reader"></param>
            /// <returns></returns>
            protected override bool LoadFromRecordset(DBReader reader)
            {
                m_SantoPatrono = reader.Read("SantoPatrono", m_SantoPatrono);
                m_GiornoFestivo = reader.Read("GiornoFestivo", m_GiornoFestivo);
                m_Prefisso = reader.Read("Prefisso", m_Prefisso);
                m_Sigla = reader.Read("Sigla", m_Sigla);
                return base.LoadFromRecordset(reader);
            }

            /// <summary>
            /// Salva nel db
            /// </summary>
            /// <param name="writer"></param>
            /// <returns></returns>
            protected override bool SaveToRecordset(DBWriter writer)
            {
                writer.Write("SantoPatrono", m_SantoPatrono);
                writer.Write("GiornoFestivo", m_GiornoFestivo);
                writer.Write("Prefisso", m_Prefisso);
                writer.Write("Sigla", m_Sigla);
                return base.SaveToRecordset(writer);
            }

            /// <summary>
            /// Prepara lo schema
            /// </summary>
            /// <param name="table"></param>
            protected override void PrepareDBSchemaFields(DBTable table)
            {
                base.PrepareDBSchemaFields(table);

                var c = table.Fields.Ensure("SantoPatrono", typeof(string), 255);
                c = table.Fields.Ensure("GiornoFestivo", typeof(string), 255);
                c = table.Fields.Ensure("Prefisso", typeof(string), 255);
                c = table.Fields.Ensure("Sigla", typeof(string), 255);
            }

            /// <summary>
            /// Prepara i vincoli
            /// </summary>
            /// <param name="table"></param>
            protected override void PrepareDBSchemaConstraints(DBTable table)
            {
                base.PrepareDBSchemaConstraints(table);

                var c = table.Constraints.Ensure("idxSanto", new string[] { "SantoPatrono", "GiornoFestivo" }, DBFieldConstraintFlags.None);
                c = table.Constraints.Ensure("idxPrefisso", new string[] { "Sigla", "Prefisso" }, DBFieldConstraintFlags.None);
            }

            /// <summary>
            /// Serializzazione xml
            /// </summary>
            /// <param name="writer"></param>
            protected override void XMLSerialize(XMLWriter writer)
            {
                writer.WriteAttribute("SantoPatrono", m_SantoPatrono);
                writer.WriteAttribute("GiornoFestivo", m_GiornoFestivo);
                writer.WriteAttribute("Prefisso", m_Prefisso);
                writer.WriteAttribute("Sigla", m_Sigla);
                base.XMLSerialize(writer);
            }

            /// <summary>
            /// Deserializzazione xml
            /// </summary>
            /// <param name="fieldName"></param>
            /// <param name="fieldValue"></param>
            protected override void SetFieldInternal(string fieldName, object fieldValue)
            {
                switch (fieldName ?? "")
                {
                    case "SantoPatrono":
                        {
                            this.m_SantoPatrono = DMD.XML.Utils.Serializer.DeserializeString(fieldValue, this.m_SantoPatrono);
                            break;
                        }

                    case "GiornoFestivo":
                        {
                            this.m_GiornoFestivo = DMD.XML.Utils.Serializer.DeserializeString(fieldValue, this.m_GiornoFestivo);
                            break;
                        }

                    case "Prefisso":
                        {
                            this.m_Prefisso = DMD.XML.Utils.Serializer.DeserializeString(fieldValue, this.m_Prefisso);
                            break;
                        }

                    case "Sigla":
                        {
                            this.m_Sigla = DMD.XML.Utils.Serializer.DeserializeString(fieldValue, this.m_Prefisso);
                            break;
                        }

                    default:
                        {
                            base.SetFieldInternal(fieldName, fieldValue);
                            break;
                        }
                }
            }


          
        }
    }
}

File: MiniDOMBase/DMD/Luoghi/Nazioni/CNazione.vb
Imports minidom
Imports minidom.Sistema
Imports minidom.Databases
Imports minidom.Anagrafica

Partial Public Class Anagrafica

    <Serializable> _
    Public Class CNazione
        Inherits LuogoISTAT

        Private m_NumeroAbitanti As Integer
        Private m_NomeAbitanti As String
        Private m_SantoPatrono As String
        Private m_GiornoFestivo As String
        Private m_Prefisso As String
        Private m_Sigla As String


        Public Sub New()
        End Sub

        Public Sub New(ByVal nome As String)
            MyBase.New(nome)
        End Sub

        Public Property NumeroAbitanti As Integer
            Get
                Return Me.m_NumeroAbitanti
            End Get
            Set(value As Integer)
                If (value < 0) Then Throw New ArgumentOutOfRangeException("NumeroAbitanti")
                Dim oldValue As Integer = Me.m_NumeroAbitanti
                If (oldValue = value) Then Exit Property
                Me.m_NumeroAbitanti = value
                Me.DoChanged("NumeroAbitanti", value, oldValue)
            End Set
        End Property

        Public Property NomeAbitanti As String
            Get
                Return Me.m_NomeAbitanti
            End Get
            Set(value As String)
                value = Trim(value)
                Dim oldValue As String = Me.m_NomeAbitanti
                If (oldValue = value) Then Exit Property
                Me.m_NomeAbitanti = value
                Me.DoChanged("NomeAbitanti", value, oldValue)
            End Set
        End Property

        Public Property SantoPatrono As String
            Get
                Return Me.m_SantoPatrono
            End Get
            Set(value As String)
                value = Trim(value)
                Dim oldValue As String = Me.m_SantoPatrono
                If (oldValue = value) Then Exit Property
                Me.m_SantoPatrono = value
                Me.DoChanged("SantoPatrono", value, oldValue)
            End Set
        End Property

        Public Property GiornoFestivo As String
            Get
                Return Me.m_GiornoFestivo
            End Get
            Set(value As String)
                value = Trim(value)
                Dim oldValue As String = Me.m_GiornoFestivo
                If (oldValue = value) Then Exit Property
                Me.m_GiornoFestivo = value
                Me.DoChanged("GiornoFestivo", value, oldValue)
            End Set
        End Property

        Public Property Prefisso As String
            Get
                Return Me.m_Prefisso
            End Get
            Set(value As String)
                value = Formats.ParsePhoneNumber(value)
                Dim oldValue As String = Me.m_Prefisso
                If (oldValue = value) Then Exit Property
                Me.m_Prefisso = value
                Me.DoChanged("Prefisso", value, oldValue)
            End Set
        End Property

        Public Property Sigla As String
            Get
                Return Me.m_Sigla
            End Get
            Set(value As String)
                value = Trim(value)
                Dim oldValue As String = Me.m_Sigla
                If (oldValue = value) Then Exit Property
                Me.m_Sigla = value
                Me.DoChanged("Sigla", value, oldValue)
            End Set
        End Property

        Protected Friend Overrides Function GetConnection() As Databases.CDBConnection
            Return Anagrafica.Luoghi.Database
        End Function

        Public Overrides Function GetModule() As CModule
            Return Luoghi.Nazioni.Module
        End Function

        Public Overrides Function GetTableName() As String
            Return "tbl_Luoghi_Nazioni"
        End Function

        Protected Overrides Function LoadFromRecordset(reader As DBReader) As Boolean
            Me.m_NumeroAbitanti = reader.Read("NumeroAbitanti", Me.m_NumeroAbitanti)
            Me.m_NomeAbitanti = reader.Read("NomeAbitanti", Me.m_NomeAbitanti)
            Me.m_SantoPatrono = reader.Read("SantoPatrono", Me.m_SantoPatrono)
            Me.m_GiornoFestivo = reader.Read("GiornoFestivo", Me.m_GiornoFestivo)
            Me.m_Prefisso = reader.Read("Prefisso", Me.m_Prefisso)
            Me.m_Sigla = reader.Read("Sigla", Me.m_Sigla)
            Return MyBase.LoadFromRecordset(reader)
        End Function

        Protected Overrides Function SaveToRecordset(writer As DBWriter) As Boolean
            writer.Write("NumeroAbitanti", Me.m_NumeroAbitanti)
            writer.Write("NomeAbitanti", Me.m_NomeAbitanti)
            writer.Write("SantoPatrono", Me.m_SantoPatrono)
            writer.Write("GiornoFestivo", Me.m_GiornoFestivo)
            writer.Write("Prefisso", Me.m_Prefisso)
            writer.Write("Sigla", Me.m_Sigla)
            Return MyBase.SaveToRecordset(writer)
        End Function

        Protected Overrides Sub XMLSerialize(writer As XML.XMLWriter)
            writer.WriteAttribute("NumeroAbitanti", Me.m_NumeroAbitanti)
            writer.WriteAttribute("NomeAbitanti", Me.m_NomeAbitanti)
            writer.WriteAttribute("SantoPatrono", Me.m_SantoPatrono)
            writer.WriteAttribute("GiornoFestivo", Me.m_GiornoFestivo)
            writer.WriteAttribute("Prefisso", Me.m_Prefisso)
            writer.WriteAttribute("Sigla", Me.m_Sigla)
            MyBase.XMLSerialize(writer)
        End Sub

        Protected Overrides Sub SetFieldInternal(fieldName As String, fieldValue As Object)
            Select Case fieldName
                Case "NumeroAbitanti" : XML.Utils.Serializer.Read(Me.m_NumeroAbitanti, fieldValue)
                Case "NomeAbitanti" : XML.Utils.Serializer.Read(Me.m_NomeAbitanti, fieldValue)
                Case "SantoPatrono" : XML.Utils.Serializer.Read(Me.m_SantoPatrono, fieldValue)
                Case "GiornoFestivo" : XML.Utils.Serializer.Read(Me.m_GiornoFestivo, fieldValue)
                Case "Prefisso" : XML.Utils.Serializer.Read(Me.m_Prefisso, fieldValue)
                Case "Sigla" : XML.Utils.Serializer.Read(Me.m_Sigla, fieldValue)
                Case Else : MyBase.SetFieldInternal(fieldName, fieldValue)
            End Select
        End Sub

        Protected Overrides Sub OnAfterSave(e As SystemEvent)
            MyBase.OnAfterSave(e)
            Luoghi.Nazioni.UpdateCached(Me)
        End Sub
    End Class


End Class